pkg/logs/sender/sender_test.go

```diff
@@ -87,3 +139,79 @@ func TestSenderNotBlockedByAdditional(t *testing.T) {
 	sender.Stop()
 	destinationsCtx.Stop()
 }
+
+func TestDualShipEndpoints(t *testing.T) {
+	source := config.NewLogSource("", &config.LogsConfig{})
+
+	input := make(chan *message.Message, 1)
+	mainInput := make(chan *message.Message, 1)
+	backupInput := make(chan *message.Message, 1)
+	mainOutput := make(chan *message.Message)
+	backupOutput := make(chan *message.Message)
+
+	mainDest := newMockDestination()
+	mainDests := client.NewDestinations(mainDest, []client.Destination{})
+	backupDest := newMockDestination()
+	backupDests := client.NewDestinations(backupDest, []client.Destination{})
+
+	mainMockStrategy := newMockStrategy()
+	backupMockStrategy := newMockStrategy()
+
+	mainSender := NewSender(mainInput, mainOutput, mainDests, mainMockStrategy)
+	backupSender := NewSender(backupInput, backupOutput, backupDests, backupMockStrategy)
+
+	SplitSenders(input, mainSender, backupSender)
+
+	mainSender.Start()
+	backupSender.Start()
+
+	// Scenario 1: Both senders fail, and then both recover
+
+	// Test both output's get the line
+	input <- newMessage([]byte("fake line"), source, "")
+	<-mainOutput
+	<-backupOutput
+
+	backupDest.errChan <- true
+
+	// Main should get the message - backup should not.
+	input <- newMessage([]byte("fake line"), source, "")
+	<-mainOutput
+	<-backupMockStrategy.sendFailed
+
+	mainDest.errChan <- true
+
+	// Both senders are in a failed state. SplitSenders will now block until
+	// one succeeds regardless of the error state.
+	input <- newMessage([]byte("fake line"), source, "")
+	<-mainMockStrategy.sendFailed
+	<-backupMockStrategy.sendFailed
+
+	mainDest.errChan <- false
+
+	// Main has recovered and should should get the message - backup should not.
+	input <- newMessage([]byte("1"), source, "")
+	<-mainOutput
+	<-backupMockStrategy.sendFailed
+
+	backupDest.errChan <- false
+
+	// Both senders are now recovered - both get the messsage
+	input <- newMessage([]byte("fake line"), source, "")
+	<-mainOutput
+	<-backupOutput
+
+	// Scenario 2: One sender fails and then recovers
+
+	mainDest.errChan <- true
+
+	input <- newMessage([]byte("fake line"), source, "")
+	<-mainMockStrategy.sendFailed
+	<-backupOutput
+
+	mainDest.errChan <- false
+
+	input <- newMessage([]byte("fake line"), source, "")
+	<-mainOutput
+	<-backupOutput
+}
```

Comment:
Still a bit hesitant about a chained test like this but it's not a blocker

Reply:
I see where you are coming from. 
I agree - It would be nice to break it up. But I think it's important to test that it recovers gracefully as it transitions between states (doesn't deadlock on a channel or something). 

I ran into a lot of issues writing this test where previous state impacted future behavior because this logic is inherently stateful (and as a result found/fixed many bugs)